File: master/buildbot/config/builder.py
# This file is part of Buildbot.  Buildbot is free software: you can
# redistribute it and/or modify it under the terms of the GNU General Public
# License as published by the Free Software Foundation, version 2.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# this program; if not, write to the Free Software Foundation, Inc., 51
# Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#
# Copyright Buildbot Team Members


from buildbot.config.checks import check_param_length
from buildbot.config.errors import error
from buildbot.db.model import Model
from buildbot.util import bytes2unicode
from buildbot.util import config as util_config
from buildbot.util import safeTranslate

RESERVED_UNDERSCORE_NAMES = ["__Janitor"]


class BuilderConfig(util_config.ConfiguredMixin):

    def __init__(self, name=None, workername=None, workernames=None,
                 builddir=None, workerbuilddir=None, factory=None,
                 tags=None,
                 nextWorker=None, nextBuild=None, locks=None, env=None,
                 properties=None, collapseRequests=None, description=None,
                 canStartBuild=None, defaultProperties=None
                 ):
        # name is required, and can't start with '_'
        if not name or type(name) not in (bytes, str):
            error("builder's name is required")
            name = '<unknown>'
        elif name[0] == '_' and name not in RESERVED_UNDERSCORE_NAMES:
            error(f"builder names must not start with an underscore: '{name}'")
        try:
            self.name = bytes2unicode(name, encoding="ascii")
        except UnicodeDecodeError:
            error("builder names must be unicode or ASCII")

        # factory is required
        if factory is None:
            error(f"builder '{name}' has no factory")
        from buildbot.process.factory import BuildFactory
        if factory is not None and not isinstance(factory, BuildFactory):
            error(f"builder '{name}'s factory is not a BuildFactory instance")
        self.factory = factory

        # workernames can be a single worker name or a list, and should also
        # include workername, if given
        if isinstance(workernames, str):
            workernames = [workernames]
        if workernames:
            if not isinstance(workernames, list):
                error(f"builder '{name}': workernames must be a list or a string")
        else:
            workernames = []

        if workername:
            if not isinstance(workername, str):
                error(f"builder '{name}': workername must be a string but it is {repr(workername)}")
            workernames = workernames + [workername]
        if not workernames:
            error(f"builder '{name}': at least one workername is required")

        self.workernames = workernames

        # builddir defaults to name
        if builddir is None:
            builddir = safeTranslate(name)
            builddir = bytes2unicode(builddir)
        self.builddir = builddir

        # workerbuilddir defaults to builddir
        if workerbuilddir is None:
            workerbuilddir = builddir
        self.workerbuilddir = workerbuilddir

        # remainder are optional
        if tags:
            if not isinstance(tags, list):
                error(f"builder '{name}': tags must be a list")
            bad_tags = any((tag for tag in tags if not isinstance(tag, str)))
            if bad_tags:
                error(f"builder '{name}': tags list contains something that is not a string")

            if len(tags) != len(set(tags)):
                dupes = " ".join({x for x in tags if tags.count(x) > 1})
                error(f"builder '{name}': tags list contains duplicate tags: {dupes}")
        else:
            tags = []

        self.tags = tags

        self.nextWorker = nextWorker
        if nextWorker and not callable(nextWorker):
            error('nextWorker must be a callable')
        self.nextBuild = nextBuild
        if nextBuild and not callable(nextBuild):
            error('nextBuild must be a callable')
        self.canStartBuild = canStartBuild
        if canStartBuild and not callable(canStartBuild):
            error('canStartBuild must be a callable')

        self.locks = locks or []
        self.env = env or {}
        if not isinstance(self.env, dict):
            error("builder's env must be a dictionary")

        self.properties = properties or {}
        for property_name in self.properties:
            check_param_length(property_name, f'Builder {self.name} property',
                               Model.property_name_length)

        self.defaultProperties = defaultProperties or {}
        for property_name in self.defaultProperties:
            check_param_length(property_name, f'Builder {self.name} default property',
                               Model.property_name_length)

        self.collapseRequests = collapseRequests

        self.description = description

    def getConfigDict(self):
        # note: this method will disappear eventually - put your smarts in the
        # constructor!
        rv = {
            'name': self.name,
            'workernames': self.workernames,
            'factory': self.factory,
            'builddir': self.builddir,
            'workerbuilddir': self.workerbuilddir,
        }
        if self.tags:
            rv['tags'] = self.tags
        if self.nextWorker:
            rv['nextWorker'] = self.nextWorker
        if self.nextBuild:
            rv['nextBuild'] = self.nextBuild
        if self.locks:
            rv['locks'] = self.locks
        if self.env:
            rv['env'] = self.env
        if self.properties:
            rv['properties'] = self.properties
        if self.defaultProperties:
            rv['defaultProperties'] = self.defaultProperties
        if self.collapseRequests is not None:
            rv['collapseRequests'] = self.collapseRequests
        if self.description:
            rv['description'] = self.description
        return rv

File: e2e/tests/pages/builder.ts
/*
  This file is part of Buildbot.  Buildbot is free software: you can
  redistribute it and/or modify it under the terms of the GNU General Public
  License as published by the Free Software Foundation, version 2.

  This program is distributed in the hope that it will be useful, but WITHOUT
  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
  FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
  details.

  You should have received a copy of the GNU General Public License along with
  this program; if not, write to the Free Software Foundation, Inc., 51
  Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

  Copyright Buildbot Team Members
*/

import {expect, Page} from "@playwright/test";
import {BasePage} from "./base";

export class BuilderPage {
  static async gotoBuildersList(page: Page) {
    await page.goto('/#/builders');
    await BasePage.waitUntilFinishedLoading(page);
  }

  static async goto(page: Page, builder: string) {
    await BuilderPage.gotoBuildersList(page);
    await page.getByRole('link', {name: builder, exact: true}).click();
    await page.waitForURL(/\/#\/builders\/[0-9]+$/);
    await BasePage.waitUntilFinishedLoading(page);
  }

  static async gotoForce(page: Page, builder: string, forceName: string) {
    await BuilderPage.goto(page, builder);
    await page.locator(`button:text("${forceName}")`).first().click();
    await expect(page.locator(".modal-title.h4")).toBeEnabled();
  }

  static async gotoBuild(page: Page, builder: string, buildRef: string) {
    await BuilderPage.goto(page, builder);
    await BuilderPage.buildLinkByBuildRef(page, buildRef).click();
    await page.waitForURL(/\/#\/builders\/[0-9]\/builds\/[0-9]+$/);
    await BasePage.waitUntilFinishedLoading(page);
  }

  static async getLastFinishedBuildNumber(page: Page) {
    const finishedBuildCss =
      '.bb-badge-round.results_SUCCESS, ' +
      '.bb-badge-round.results_WARNINGS, ' +
      '.bb-badge-round.results_FAILURE, ' +
      '.bb-badge-round.results_SKIPPED, ' +
      '.bb-badge-round.results_EXCEPTION, ' +
      '.bb-badge-round.results_RETRY, ' +
      '.bb-badge-round.results_CANCELLED ';

    const el = page.getByTestId('build-link').and(page.locator(finishedBuildCss));
    if (await el.count() > 0) {
      const text = await el.first().textContent();
      if (text === null) {
        return 0;
      }

      const m = /.*\(([0-9]*)\)/.exec(text);
      if (m !== null) {
        return Number.parseInt(m[1]);
      }
      return Number.parseInt(text);
    }
    return 0;
  }


  static async getBuildResult(page: Page, buildNumber: number) {
    const links = BuilderPage.buildLinkByBuildRef(page, buildNumber.toString());

    const resultTypes = [
      ['.bb-badge-round.results_SUCCESS', "SUCCESS"],
      ['.bb-badge-round.results_WARNINGS', "WARNINGS"],
      ['.bb-badge-round.results_FAILURE', "FAILURE"],
      ['.bb-badge-round.results_SKIPPED', "SKIPPED"],
      ['.bb-badge-round.results_EXCEPTION', "EXCEPTION"],
      ['.bb-badge-round.results_RETRY', "RETRY"],
      ['.bb-badge-round.results_CANCELLED', "CANCELLED"]
    ];

    for (let i = 0; i < resultTypes.length; ++i) {
      const css = resultTypes[i][0];
      const resultType = resultTypes[i][1];

      if (await links.and(page.locator(css)).count() > 0) {
        return resultType;
      }
    }

    return "NOT FOUND";
  }

  static buildLinkByBuildRef(page: Page, buildRef: string) {
    return page.getByTestId("build-link").and(page.locator(`:text("${buildRef}")`));
  }

  static async waitBuildFinished(page: Page, reference: number) {
    await expect.configure({ timeout: 30000 }).poll(async () => {
      const currentBuildCount = await BuilderPage.getLastFinishedBuildNumber(page);
      return currentBuildCount === reference;
    }, {
      message: "build count has been incremented",
    }).toBeTruthy();
  }

  static async waitGoToBuild(page: Page, expectedBuildNumber: number) {
    await expect.poll(async () => {
      let buildUrl = page.url();
      const split = buildUrl.split("/");
      const buildsPart = split[split.length-2];
      const number = Number.parseInt(split[split.length-1]);
      if (buildsPart !== "builds") {
        return false;
      }
      return (number === expectedBuildNumber);
    }, {
      message: "went into build"
    }).toBeTruthy();
  }

  static async clickStopButton(page: Page) {
    await page.locator('button:text("Stop")').click();
  }

  static async clickPreviousButtonAndWait(page: Page) {
    await BasePage.waitUntilUrlChanged(page, async () => {
      await page.getByRole("link", {name: "Previous"}).click();
    });
    await BasePage.waitUntilFinishedLoading(page);
  }

  static async clickNextButtonAndWait(page: Page) {
    await BasePage.waitUntilUrlChanged(page, async () => {
      await page.getByRole("link", {name: "Next"}).click();
    });
    await BasePage.waitUntilFinishedLoading(page);
  }

  static async clickRebuildButton(page: Page) {
    await BasePage.waitUntilUrlChanged(page, async () => {
      await page.getByRole("button", {name: "Rebuild"}).click();
    });
  }

  static async checkBuilderURL(page: Page, builder: string) {
    await expect(page.locator(`a:text("${builder}")`).count()).toBeGreaterThan(0);
  }
}
